python-avd/pyavd/_cv/client/async_decorators.py

```diff
@@ -0,0 +1,144 @@
+            if list_field not in arguments:
+                msg = f"grpc_msg_size_handler decorator is unable to find the list_field '{list_field}' in the given arguments to '{func.__name__}'."
+                raise KeyError(msg)
+
+            list_value: list = arguments[list_field]
+            if not isinstance(list_value, list):
+                msg = f"grpc_msg_size_handler decorator expected the value of the list_field '{list_field}' to be a list. Got '{type(list_value)}'"
+                raise TypeError(msg)
+
+            LOGGER.info("wrapper_grpc_msg_size_handler: Called '%s' with '%s' items", func.__name__, len(list_value))
+
+            if len(list_value) < 2:
+                # No need to try/except if we cannot split the list.
+                return await func(*args, **kwargs)
+
+            try:
+                return await func(*args, **kwargs)
+            except CVMessageSizeExceeded as e:
+                # At minimum try to split in two.
+                # The double negatives make // round up instead of down.
+                ratio = max(2, -(-e.size // e.max_size))
+                chunk_size = len(list_value) // ratio
+                LOGGER.info(
+                    "wrapper_grpc_msg_size_handler: Message size %s exceeded the max of %s. Splitting into %s smaller calls with up to %s items each.",
+                    e.size,
+                    e.max_size,
+                    ratio,
+                    chunk_size,
+                )
+                # For every chunk we call ourselves recursively, so we can catch any further needs of splitting.
+                result = []
+                for chunk in batch(list_value, chunk_size):
+                    arguments[list_field] = chunk
+                    result.extend(await wrapper_grpc_msg_size_handler(*bound_arguments.args, **bound_arguments.kwargs))
+                return result
+
+        return wrapper_grpc_msg_size_handler
+
+    return decorator_grpc_msg_size_handler
+
+
+class LimitCvVersion:
+    """
+    Decorator used to limit the supported CloudVision versions for a certain method.
+
+    The decorator will maintain a map of decorated function variants and their supported versions.
+
+    The decorator will only work in CvClient class methods since it expects the _cv_client attribute on 'self'.
+    """
+
+    versioned_funcs: ClassVar[dict[str, dict[tuple[CvVersion, CvVersion], Callable]]] = {}
+    """
+    Map of versioned functions keyed by function name.
+
+    {
+        <function_name>: {
+            (<min_cv_version>, <max_cv_version>): <method>,
+            (<min_cv_version>, <max_cv_version>): <method>,
+        }
+    }
+    """
+
+    def __init__(self, min_ver: str = "2024.1.0", max_ver: str = CVAAS_VERSION_STRING) -> None:
+        """__init__ is called with the arguments of the decorator."""
+        # Storing these on the instance so it can be read in the __call__ method.
+        self.min_version = CvVersion(min_ver)
```

Comment:
CvVersion could raise (yes it is used internally but maybe this should be handled in the CvVersion class?)

```
>>> LimitCvVersion("blah", "17")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/guillaumemulocher/Github/gmuloc/ansible-avd/python-avd/pyavd/_cv/client/async_decorators.py", line 100, in __init__
    self.min_version = CvVersion(min_ver)
  File "/Users/guillaumemulocher/Github/gmuloc/ansible-avd/python-avd/pyavd/_cv/client/versioning.py", line 13, in __init__
    self.version_value = self._version_as_value()
  File "/Users/guillaumemulocher/Github/gmuloc/ansible-avd/python-avd/pyavd/_cv/client/versioning.py", line 29, in _version_as_value
    return int(f"{elements[0]:04}{elements[1]:02}{elements[2]:02}")
IndexError: list index out of range
```

Reply:
Spending extra runtime to check that the developer gave correct input seems inefficient to me.

Reply:
fixed